src/processors/transform.rs

```diff
@@ -0,0 +1,311 @@
+pub struct Init;
+impl Component for Init {
+    type Storage = NullStorage<Init>;
+}
+
+/// Transformation processor.
+/// Handles updating `Transform` components based on the `LocalTransform` component and parents.
+pub struct TransformProcessor {
+    // Map of entities to index in sorted vec.
+    indices: HashMap<Entity, usize>,
+
+    // Vec of entities with parents before children.
+    sorted: Vec<(Entity, Option<Entity>)>,
+
+    // Possible candidates for swapping parents and children in sorted.
+    swap_candidates: Vec<(Entity, Entity)>,
+
+    // New entities in current update
+    new: Vec<Entity>,
+}
+
+impl TransformProcessor {
+    pub fn new() -> TransformProcessor {
+        TransformProcessor {
+            indices: HashMap::new(),
+            sorted: Vec::new(),
+            swap_candidates: Vec::new(),
+            new: Vec::new(),
+        }
+    }
+}
+
+impl Processor<Arc<Mutex<Context>>> for TransformProcessor {
+    fn run(&mut self, arg: RunArg, _: Arc<Mutex<Context>>) {
+        let (entities, mut locals, mut globals, mut init) = arg.fetch(|w| {
+            (w.entities(), w.write::<LocalTransform>(), w.write::<Transform>(), w.write::<Init>())
+        });
+
+        for (entity, local, _) in (&entities, &locals, !&init).iter() {
+            self.indices.insert(entity, self.sorted.len());
+            self.sorted.push((entity, local.parent));
+            self.new.push(entity.clone());
+        }
+
+        for entity in self.new.drain(..) {
+            init.insert(entity, Init);
+        }
+
+        // Compute transforms (global) from local transforms and parents.
+        for &(entity, parent_option) in self.sorted.iter() {
+            let mut local = locals.get_mut(entity).unwrap();
+            if local.dirty {
+                let combined_transform = match parent_option {
+                    Some(parent) => {
+                        self.swap_candidates.push((entity, parent));
+
+                        if let Some(parent_global) = globals.get(parent) {
+                            Matrix4::from(parent_global.0) * Matrix4::from(local.matrix())
+                        } else {
+                            Matrix4::from(local.matrix())
+                        }
+                    }
+                    None => Matrix4::from(local.matrix()),
+                };
+
+                if let Some(global) = globals.get_mut(entity) {
+                    global.0 = combined_transform.into();
+                }
+
+                local.dirty = false;
+            }
+        }
+
+        // Checks whether the child is before the parent.
+        // If so, it swaps their positions.
+        for (entity, parent) in self.swap_candidates.drain(..) {
+            let parent_index: usize = self.indices.get(&parent).unwrap().clone();
+            let index: usize = self.indices.get(&entity).unwrap().clone();
+            if parent_index > index {
+                self.sorted.swap(parent_index.clone(), index);
```

Comment:
This is done after the world transform calculus. Does that mean we'd have a one frame of invalid world transforms in case such a reverse dependency is injected?